tools/clang/lib/CodeGen/CGHLSLMS.cpp

```diff
@@ -3190,182 +3190,190 @@ static void AddOpcodeParamForIntrinsics(HLModule &HLM
+      Value *Vec1 = UndefValue::get(ToTy);
+      return Builder.CreateInsertElement(Vec1, V, (uint64_t)0);
+    } else if (FromTy->isVectorTy() && vecSize == 1) {
+      Value *V = Builder.CreateLoad(Ptr);
+      // VectorTrunc
+      // Change vector into vec1.
+      return Builder.CreateShuffleVector(V, V, {0});
+    } else if (FromTy->isArrayTy()) {
+      llvm::Type *FromEltTy = FromTy->getArrayElementType();
+
+      llvm::Type *ToEltTy = ToTy->getVectorElementType();
+      if (FromTy->getArrayNumElements() == vecSize && FromEltTy == ToEltTy) {
+        // ArrayToVector.
+        Value *NewLd = UndefValue::get(ToTy);
+        Value *zeroIdx = Builder.getInt32(0);
+        for (unsigned i = 0; i < vecSize; i++) {
+          Value *GEP = Builder.CreateInBoundsGEP(
+              Ptr, {zeroIdx, Builder.getInt32(i)});
+          Value *Elt = Builder.CreateLoad(GEP);
+          NewLd = Builder.CreateInsertElement(NewLd, Elt, i);
+        }
+        return NewLd;
+      }
+    }
+  } else if (FromTy == Builder.getInt1Ty()) {
+    Value *V = Builder.CreateLoad(Ptr);
+    // BoolCast
+    DXASSERT_NOMSG(ToTy->isIntegerTy());
+    return Builder.CreateZExt(V, ToTy);
+  }
+
+  return UndefValue::get(ToTy);
+}
 
-      Builder.CreateStore(Val, Ptr);
-      deadInsts.emplace_back(stInst);
+static Value  *CastStValue(Value *Ptr, Value *V, llvm::Type *FromTy, llvm::Type *ToTy, IRBuilder<> &Builder) {
+  if (ToTy->isVectorTy()) {
+    unsigned vecSize = ToTy->getVectorNumElements();
+    if (vecSize == 1 && ToTy->getVectorElementType() == FromTy) {
+      // ScalarToVec1Splat
+      // Change vec1 back to scalar.
+      Value *Elt = Builder.CreateExtractElement(V, (uint64_t)0);
+      return Elt;
+    } else if (FromTy->isVectorTy() && vecSize == 1) {
+      // VectorTrunc
+      // Change vec1 into vector.
+      // Should not happen.
+      // Reported error at Sema::ImpCastExprToType.
+      DXASSERT_NOMSG(0);
+    } else if (FromTy->isArrayTy()) {
+      llvm::Type *FromEltTy = FromTy->getArrayElementType();
+
+      llvm::Type *ToEltTy = ToTy->getVectorElementType();
+      if (FromTy->getArrayNumElements() == vecSize && FromEltTy == ToEltTy) {
+        // ArrayToVector.
+        Value *zeroIdx = Builder.getInt32(0);
+        for (unsigned i = 0; i < vecSize; i++) {
+          Value *Elt = Builder.CreateExtractElement(V, i);
+          Value *GEP = Builder.CreateInBoundsGEP(
+              Ptr, {zeroIdx, Builder.getInt32(i)});
+          Builder.CreateStore(Elt, GEP);
+        }
+        // The store already done.
+        // Return undef to ignore use of the return value.
+        return UndefValue::get(ToTy);
+      }
     }
+  } else if (FromTy == Builder.getInt1Ty()) {
+    // BoolCast
+    // Change i1 to ToTy.
+    DXASSERT_NOMSG(ToTy->isIntegerTy());
+    Value *CastV = Builder.CreateICmpNE(V, ConstantInt::get(V->getType(), 0));
+    return CastV;
   }
-  deadInsts.emplace_back(BCI);
+
+  return UndefValue::get(ToTy);
```

Comment:
Should we return nullptr if the value was not translated? If this is an error condition then we should assert here.